### ISSUE TITLE
Issue reproducing the CodeXGLUE CSNet text-to-code results

### ISSUE THREAD

**Issue by eelxpeng** [rizwan09/REDCODER#3]
@rizwan09 I'm trying to reproduce CodeXGLUE CSNet python code generation results in Table 2. However, I cannot get the correct results as reported in the paper.

Here are the steps I followed (mainly categrized into three major steps, download data, preprare retrievals and training generator). Environment setup largely follows the READMEs and omitted here.

**Step 0**: download checkpoint and data file, and put them under `$DATA_ROOT/data/`
1. Download SCODE-R checkpoint from [link provided](https://drive.google.com/drive/folders/1qrt6DkJKp43TNN_wCTVCP98BOtD-9o9J?usp=sharing), and specifically "DPR_TEXT_CODE_PYTHON.cp" file.
2. Download retrieval database file from the same link, i.e. retrieval_database/codexglue_csnet/python_dedupe_definitions_v2.pkl
3. Download graphcodebert model from the [huggingface link](https://huggingface.co/microsoft/graphcodebert-base)
4. Download csnet python train/valid/test data from [google drive](https://drive.google.com/file/d/1m1IvGgPhDBg-SL-LajtFGTLyAJVbD0i3) shared in the [PLBART repo script](https://github.com/wasiahmad/PLBART/blob/main/data/codeXglue/download.sh) under $DATA_ROOT/data/code-to-text/python
5. Download plbart checkpoint from the [link](https://drive.google.com/file/d/19OLKx0YY0yVorzZa-caFW0-hALVvX7gt) in the SCODE-G/download_plbart.sh script.

**Step 1**: Prepare retrievals and save data under `$DATA_ROOT/output/`
1. Generate encodings for retrieval database candidates:
Since the checkpoint contains `pretrained_model_cfg` pointing to wrong graphcodebert path, I had to modify `SCODE-R/dpr/utils/options.py` line 120 to add a condition `and param != "pretrained_model_cfg"` to not override the cfg from the checkpoint.
```
CHECKPOINT="${DATA_ROOT}/data/DPR_TEXT_CODE_PYTHON.cp"
CANDIDATE_FILE_NAME="python_dedupe_definitions_v2.pkl"
CANDIDATE_FILE="${DATA_ROOT}/data/retrieval_database/codexglue_csnet/${CANDIDATE_FILE_NAME}"
DEVICES="0,1"
NUM_DEVICES=2
ENCODDING_CANDIDATE_PREFIX="${DATA_ROOT}/output/encoddings_${CANDIDATE_FILE_NAME}}"
PRETRAINED_MODEL_PATH="${DATA_ROOT}/data/graphcodebert-base"
          

CUDA_VISIBLE_DEVICES=${DEVICES} python  -m torch.distributed.launch \
          --nproc_per_node=${NUM_DEVICES} generate_dense_embeddings.py \
          --model_file  ${CHECKPOINT}  \
          --encoder_model_type hf_roberta \
          --pretrained_model_cfg  ${PRETRAINED_MODEL_PATH} \
          --batch_size 512 \
          --ctx_file  ${CANDIDATE_FILE} \
          --shard_id 0 \
          --num_shards 1 \
          --out_file  ${ENCODDING_CANDIDATE_PREFIX}
```

2. Doing retrieval for csnet python text-to-code training/valid/test data
```
for split in train valid test; do
	TOP_K=20
	RETRIEVAL_RESULT_FILE="${DATA_ROOT}/output/python_csnet_text_code_retrieval_${split}_20.json"
	CHECKPOINT="${DATA_ROOT}/data/DPR_TEXT_CODE_PYTHON.cp"
	CANDIDATE_FILE_NAME="python_dedupe_definitions_v2.pkl"
	CANDIDATE_FILE="${DATA_ROOT}/data/retrieval_database/codexglue_csnet/${CANDIDATE_FILE_NAME}"
	ENCODDING_CANDIDATE_PREFIX="${DATA_ROOT}/output/encoddings_${CANDIDATE_FILE_NAME}"
	PRETRAINED_MODEL_PATH="${DATA_ROOT}/data/graphcodebert-base"
	FILE_FOR_WHICH_TO_RETIRVE="${DATA_ROOT}/data/code-to-text/python/${split}.jsonl"

	CUDA_VISIBLE_DEVICES=0 python dense_retriever.py \
	        --model_file ${CHECKPOINT} \
	        --ctx_file  ${CANDIDATE_FILE} \
	        --qa_file ${FILE_FOR_WHICH_TO_RETIRVE} \
	        --encoded_ctx_file ${ENCODDING_CANDIDATE_PREFIX} \
	        --out_file  ${RETRIEVAL_RESULT_FILE} \
	        --n-docs  ${TOP_K} \
	        --sequence_length 256 \
	        --save_or_load_index \
	        --pretrained_model_cfg ${PRETRAINED_MODEL_PATH}
done
```
Do the similar script for valid.jsonl and test.jsonl. It results in files with retrievals:
```
python_csnet_text_code_retrieval_train_20.json
python_csnet_text_code_retrieval_valid_20.json
python_csnet_text_code_retrieval_test_20.json
```

**Step 2**: Train Generator
```
cd SCODE-G/text_to_code
```
1. data preprocessing
Since above the generated filenames are "python_csnet_text_code_retrieval_{train|valid|test}_20.json", I have to revise `SCODE-G/text_to_code/prepare_csnet_redcoder.sh` line 24 to this file pattern.
```
LANG=python
top_k=5
WITH_OR_WITHOUT_REF=no
RETDIR="${DATA_ROOT}/output"
SAVE_DIR="${DATA_ROOT}/tokenized"

bash prepare_csnet_redcoder.sh ${LANG} ${top_k} ${WITH_OR_WITHOUT_REF} ${RETDIR} ${SAVE_DIR}
```

2. start training
```
GPU=0
LANG=python
path_2_data="${DATA_ROOT}/tokenized"
PRETRAIN="${DATA_ROOT}/data/checkpoint_11_100000.pt"
SAVE_DIR="${DATA_ROOT}/checkpoint"
UPDATE_FREQ=2
BATCH_SIZE=16

bash run_csnet.sh ${GPU} ${LANG} ${path_2_data} ${PRETRAIN} ${SAVE_DIR} ${UPDATE_FREQ} ${BATCH_SIZE}
```

After running 8 epochs, the training is complete. However, the results the console reported is
```
BLEU:           8.31
Acc:            0.27
Ngram match:    8.31
Weighted ngram: 9.55
Syntax match:   26.64
Dataflow match: 22.88
CodeBLEU score: 16.85
```
This result is largely different from the results in Table 2. I'm wondering which step I did wrong. Could you help to look into the scripts and give me a hint?
Thanks!

**Comment by rizwan09** [comment]
Hi @eelxpeng,
Thank you for using REDOCDER.

Regarding python_csnet_text_code_retrieval_{train|valid|test}_20.json could you please see what is the retrieval accuracy? To make sure: as from Fig 6 in the paper, the recall at 10 should be 35+. 

But anyway I am uploading the retrieval output myself (needs ~4 hours) , so you only need to do the SCODE-G (PLBART) part. The files will be available at [LINK](https://drive.google.com/drive/folders/1qrt6DkJKp43TNN_wCTVCP98BOtD-9o9J?usp=sharing).


Please make sure the data preprocessing of SCODE-G is correct. Manually dig into the directory. Seems like it's not using any retrieved data as the performance is very similar to just using PLBART.

**Comment by eelxpeng** [comment]
@rizwan09 thanks for the quick reply! At step 1 generating retrievals, the console output `Validation results: top k documents hits accuracy` are all `1.0`. I checked the output files, top 1 retrievals seem to be exactly same as the target answer. I haven't found a straightforward way to compute recall@K.

I manually checked the preprocessing output, i.e. train.source or test.source, and the retrievals are there, right after each question separated with `_CODE_SEP_` token.

**Comment by rizwan09** [comment]
The retrieval results in the paper were reported on the test set. Wait (~30 min left) for the link and then you can try using them.

**Comment by rizwan09** [comment]
The retriever outputs are uploaded! please check now! Thanks.

**Comment by eelxpeng** [comment]
@rizwan09 thanks a lot! I'm currently running training generator part with your retrieval files. Will update once finished.

**Comment by eelxpeng** [comment]
@rizwan09 I have verified the your released retrieval files, and yes, they work as expected and matches Table 2. Thank you very much! Now I'm trying to see why my generated retrievals in Step 1 are wrong. I evaluated the retrieval performance of your released retrieval files and the one I generated in Step 1, and yours is Precision@1 19.2% and mine is Precision@1 35.2%. The script I used to evaluate is:
```
def main():
    import argparse
    parser = argparse.ArgumentParser(description='Evaluate leaderboard predictions for POJ-104 dataset.')
    parser.add_argument('--retrieval_output_file', '-r', help="filename of the labels, in txt format.")
    args = parser.parse_args()

    data = list(json.load(open(args.retrieval_output_file)))
    hits = []
    for d in data:
        hit = []
        for i, ctx in enumerate(d["ctxs"]):
            retrieval_code = ctx["text"].strip().split('_NL_')[0].strip()
            if d["answers"] == retrieval_code:
                hit.append(1)
            else:
                hit.append(0)
        hits.append(hit)
    hits = np.array(hits)

    # compute metrics
    cum_hits = np.cumsum(hits, axis=1)
    precision = cum_hits.sum(0) / (np.arange(1, hits.shape[1] + 1) * len(hits))

    print(precision)
```
It seems that my generated file in Step1 has even higher retrieval performance. However, I do noticed that, in your generated retrievals it contains `_NL_` token, that means you are using `dense_retriever_with_comments.py` to do the retrieval, instead of `dense_retriever.py`. Do you think this might be the main reason since I'm using your released dpr checkpoint?

**Comment by eelxpeng** [comment]
But I checked `dense_retriever_with_comments.py` and `dense_retriever.py` has very slight difference, and since I use `prepare_csnet_redcoder.sh` that without comments in generator data preprocessing, there should be no difference. Could you help me check if my commands in Step 1 has any issues that might mismatch your provided checkpoints, for example `--encoder_model_type hf_roberta` or `--pretrained_model_cfg  ${DATA_ROOT}/data/graphcodebert-base`?

**Comment by rizwan09** [comment]
Hi @eelxpeng, 
Sorry, was passing very bz days. Is your problem fixed? 

Regarding SCODE-R  training: as I mentioned in the readme of SCODE-R, for both w/ and w/o comments we never use any text/comment info. They are only used in SCODE-G. Both SCODE-R models are purely code retrieval (for text-to-code). 
Now as for the respective SCODE-G, you can use only the code and code plus comment. To save them (I mean for inference) so you can use either ``dense_retriever.py`` or ``dense_retriever_with_comments.py``. Their difference is line 273. 
But in general ``dense_retriever_with_comments.py`` is the one that you can use for all of them which will infact save both the code and the text/cooment. And you can just use code or code plus comment from it in SCODE-G as per your redcoder or redcoder-ext settings. 

So, the retrieval output file I released is actually using  ``dense_retriever_with_comments.py`` and both code and code plus comments info are available to use as per your settings. 

Hope this clarifies.

**Comment by eelxpeng** [comment]
Hi @rizwan09 thanks for the reply. No I haven't resolved the problem. Your shared retrieval output is good. The issue is how to correctly generate your shared retrieval output. I tried using your shared dpr checkpoint, i.e. SCODE_R_TEXT_CODE_PYTHON.cp, and I also tried training dpr with the steps in the README in SCODE-R, i.e.
```
DEVICES=0,1,2,3
NUM_DEVICES=4
PRETRAINED_MODEL_PATH="${DATA_ROOT}/data/graphcodebert-base"
TRAIN_FILE_PATH="${DATA_ROOT}/data/code-to-text/python/train.jsonl"
DEV_FILE_PATH="${DATA_ROOT}/data/code-to-text/python/valid.jsonl"
OUTPUT_DIR="${DATA_ROOT}/checkpoint-retriever"

CUDA_VISIBLE_DEVICES=${DEVICES} python -m torch.distributed.launch 
     --nproc_per_node=${NUM_DEVICES} train_dense_encoder.py 
     --max_grad_norm 2.0 
     --encoder_model_type hf_roberta 
     --pretrained_model_cfg ${PRETRAINED_MODEL_PATH}
     --eval_per_epoch 1 
     --seed 12345 
     --sequence_length 256 
     --warmup_steps 1237 
     --batch_size 8 
     --train_file ${TRAIN_FILE_PATH}
     --dev_file ${DEV_FILE_PATH}
     --output_dir ${OUTPUT_DIR}
     --learning_rate 2e-5 
     --num_train_epochs 15 
     --dev_batch_size 64 
     --val_av_rank_start_epoch 0 
     --text_to_code 
     --fp16 
```
But neither of the two checkpoints gives me good retrieval output, which can get good BLEU score when used in generator training.

To debug the issue, if not too much trouble could you help give the correct commands with your shared checkpoint SCODE_R_TEXT_CODE_PYTHON.cp in [link provided](https://drive.google.com/drive/folders/1qrt6DkJKp43TNN_wCTVCP98BOtD-9o9J?usp=sharing) to generate the retrieval output in the same way as how you generate `retriever_output_codexglue_csnet_text_to_code.zip` in the same link? That could help a lot. Thanks!

**Comment by eelxpeng** [comment]
I have also noticed that for `train_dense_encoder.py` and `train_dense_encoder_with_comments.py`, the difference is that the ctx file `python_dedupe_definitions_v2.pkl`, the former target is using `function_tokens` while the latter target uses `function`, which includes function + docstring. Which one are you using for your checkpoint? If the latter, then why `generate_dense_embeddings.py` only uses `function_tokens` and not `functions` in Line 123?

**Comment by eelxpeng** [comment]
@rizwan09 did you get a chance to see what is the issue here? Could you help give the correct retriever training commands that could lead to your shared checkpoint SCODE_R_TEXT_CODE_PYTHON.cp in [link you released](https://drive.google.com/drive/folders/1qrt6DkJKp43TNN_wCTVCP98BOtD-9o9J?usp=sharing)?
